src/Util/ModuleIterator.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace SilverStripe\Upgrader\Util;
+
+use Generator;
+use IteratorAggregate;
+use Traversable;
+
+class ModuleIterator implements IteratorAggregate
+{
+    /**
+     * @var string
+     */
+    protected $path = null;
+
+    /**
+     * @param string $path
+     */
+    public function __construct($path)
+    {
+        $this->path = $path;
+    }
+
+    /**
+     * Files that signify a module should be included
+     *
+     * @var array
+     */
+    protected $files = [
+        '_config',
+        '_config.php',
+        ConfigFile::NAME,
+    ];
+
+    /**
+     * Iterate over all module paths
+     *
+     * @param string $rootPath
+     * @return Generator
+     */
+    public function find($rootPath)
+    {
+        // Check all root modules
+        foreach (glob("{$rootPath}/*", GLOB_ONLYDIR) as $path) {
+            // Mysite base
+            if (dirname($path) === 'mysite' || $this->isModule($path)) {
+                yield $path;
+            }
+        }
+
+        // Check vendor modules
+        foreach (glob("{$rootPath}/vendor/*/*", GLOB_ONLYDIR) as $path) {
+            if ($this->isModule($path)) {
+                yield $path;
+            }
+        }
```

Comment:
Would module order matter for this case?

Reply:
I'm not sure it matters really. It all gets merged together in the end.

Reply:
I guess it was order-sensitive commands it could hit hard (something like shifting asset files around?) which we aren't doing here anyway.